### DISCUSSION TITLE
Python Version update and CPLEX installation

### DISCUSSION THREAD

**Discussion by Lakes-lead** [GRIDAPPSD/gridappsd-forum#102] [Q&A]
I tried this code in the restoration application

root@1b762c641f24:/usr/src/gridappsd-restoration# cd /opt/ibm/ILOG/CPLEX_Studio129/cplex/python/3.6/x86-64_linux/ ; python setup.py install

I keep getting this feedback, it looks like my python version is 3.4.2 i tried updating the version to 3.6 from the terminal but it isn't working any suggestions on what i can do?
Thank you
![error messgae](https://user-images.githubusercontent.com/26089614/110375777-6662ec80-8020-11eb-8111-c390fe09f6b1.PNG)

**Comment by poorva1209** [comment]
@shpoudel Please answer here for the above question. Create an issue if needed.

**Comment by shpoudel** [comment]
Hello

Thanks for the interest in the restoration application. Do you have CPLEX in your VM? There are two commands that you are trying to execute. One is cd ... and another is python setup.py install. Since your first command is not working because of the CPLEX issue, your python command is not successful. Please see step 2 in the following link where we copy the CPLEX from the local machine to docker:

https://github.com/shpoudel/WSU-Restoration

Also, there is a way to run the application without CPLEX. Please refer to the readthedocs- https://wsu-restoration.readthedocs.io/en/latest/using_gridappsd/index.html

**Reply by Lakes-lead** [reply]
@shpoudel Okay Thank you. 

I don't have CPLEX on my local machine nor my VM. 

For clarifications should I install CPLEX on my local machine(from windows terminal) or the virtual machine( using the Ubuntu terminal), if yes what are the steps involved I have executed the step 2 from the link  already but same issue.  

![CPLEX](https://user-images.githubusercontent.com/26089614/110381894-20118b80-8028-11eb-8e98-788f69bb3d82.PNG)



https://github.com/shpoudel/WSU-Restoration

**Reply by Lakes-lead** [reply]
@shpoudel is it possible to have a one on one session with you on this.
Thank you

**Reply by shpoudel** [reply]
Yes, you will need to install CPLEX in your VM if you want to run the application as it is. As I mentioned earlier,  there is a way to run the application without CPLEX. Please refer to the readthedocs- https://wsu-restoration.readthedocs.io/en/latest/using_gridappsd/index.html

Could you please identify yourself before we have a one-on-one session?

**Reply by Lakes-lead** [reply]
I am Olalekan Ogundairo , a PhD student.

**Reply by shpoudel** [reply]
Hi Olalekan. I suggest you run this application without CPLEX following the link I mentioned. Let me know how that goes and we can certainly talk more once you start on this.

**Reply by Lakes-lead** [reply]
@shpoudel i sent you a Linkedin request. Thank you.

I have tried to run it without the CPLEX, but i still get stuck at the point  in the file attached..
![Stop_point](https://user-images.githubusercontent.com/26089614/110500779-8dc0c480-80c7-11eb-99f8-ef4748489e6f.JPG)

**Reply by shpoudel** [reply]
Hi, 

When you are running without CPLEX, you don't need to do the installation for it. Please avoid the following command as you are not using CPLEX:

cd /opt/ibm/ILOG/CPLEX_Studio129/cplex/python/3.6/x86-64_linux/ ; python setup.py install

**Reply by AAndersn** [reply]
Hi, also you should double-check that the CPLEX option is commented out in lines 347 (or so) of your cloned copy of restoration_WSU.py (https://github.com/shpoudel/WSU-Restoration/blob/master/Restoration/restoration_WSU.py). Then uncomment the default python solver. It should look something like this

```
        prob.solve()
        # prob.solve(CPLEX(msg=1))
        prob.writeLP("Check.lp")
```

After that, you can copy the updated python script into the docker container using the `docker cp` command:

```
cd WSU_Restoration/
cd Restoration/
cat restoration_WSU.py
```
=> Check "prob.solve(CPLEX(msg=1))" is commented out

`docker cp restoration_WSU.py gridappsddocker_wsu_restoration_1:/usr/src/gridappsd-restoration/Restoration
`

Hope this helps!

**Reply by shpoudel** [reply]
Thanks, @AAndersn. 

Olalekan, please follow these steps once and let us know how it goes. Thanks again for your interest.

**Reply by Lakes-lead** [reply]
@shpoudel @AAndersn Thanks i have been able to run it without CPLEX installation

@shpoudel 
I still have a few questions
1) The paper publication talks about the IEEE8500/ the documentation is on test9500new, which exactly is appropriate to run. 
2) I saved the event as a .json  file and uploaded it . Do i still need to click the radio button for the fault before running the simulation or the file will create it by itself when running.    
3) For the Simulation start time do i start at exact time of event occurence or before the event occurrence, so the fault can be seen.

Thank you

**Reply by shpoudel** [reply]
@Lakes-lead Perfect!

The paper talks about the modified IEEE 8500-node which is the test9500new. In the publications, we can't say 9500-node as it is not an official IEEE feeder. Please select test9500new for the simulation

When you uploaded a json file, there is no need to click the radio button for the fault. The uploaded json file will be a trigger for the application. 

For the simulation start time, please keep it at least 1 min before the event occurrence, so that you can see the fault and application being triggered. Great question.

Keep in mind, the event file has object mrid that might have changed when the database gets updated. Please verify it by running the query. Once you start the containers, go to http://localhost:8889/bigdata/#query and upload the following query to see the switch mrid to have it in the event JSON file.


PREFIX r:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX c:  <http://iec.ch/TC57/CIM100#>
SELECT ?cimtype ?name ?bus1 ?bus2 ?id (group_concat(distinct ?phs;separator="") as ?phases) WHERE {
  SELECT ?cimtype ?name ?bus1 ?bus2 ?phs ?id WHERE {
 VALUES ?fdrid {"_AAE94E4A-2465-6F5E-37B1-3E72183A4E44"}  # 9500 node
 VALUES ?cimraw {c:LoadBreakSwitch c:Recloser c:Breaker}
 ?fdr c:IdentifiedObject.mRID ?fdrid.
 ?s r:type ?cimraw.
  bind(strafter(str(?cimraw),"#") as ?cimtype)
 ?s c:Equipment.EquipmentContainer ?fdr.
 ?s c:IdentifiedObject.name ?name.
 ?s c:IdentifiedObject.mRID ?id.
 ?t1 c:Terminal.ConductingEquipment ?s.
 ?t1 c:ACDCTerminal.sequenceNumber "1".
 ?t1 c:Terminal.ConnectivityNode ?cn1. 
 ?cn1 c:IdentifiedObject.name ?bus1.
 ?t2 c:Terminal.ConductingEquipment ?s.
 ?t2 c:ACDCTerminal.sequenceNumber "2".
 ?t2 c:Terminal.ConnectivityNode ?cn2. 
 ?cn2 c:IdentifiedObject.name ?bus2
	OPTIONAL {?swp c:SwitchPhase.Switch ?s.
 	?swp c:SwitchPhase.phaseSide1 ?phsraw.
   	bind(strafter(str(?phsraw),"SinglePhaseKind.") as ?phs) }
 } ORDER BY ?name ?phs
}
GROUP BY ?cimtype ?name ?bus1 ?bus2 ?id
ORDER BY ?cimtype ?name

**Reply by Lakes-lead** [reply]
Hello @shpoudel Thank you for your assistance so far, i am yet to still see the voltage violations are attached are the images from the gridappsd and the query execution.
![GridAppsd](https://user-images.githubusercontent.com/26089614/110736660-1049a100-81fa-11eb-8ed3-e7fc3dcc84ee.PNG)
![QUERY MESSAGE](https://user-images.githubusercontent.com/26089614/110736672-1475be80-81fa-11eb-98e1-58a9f1ae4a68.PNG)

**Reply by shpoudel** [reply]
Hello, @Lakes-lead I am not sure why the query was not copied correctly last time. Sorry about that. Once you run the query, make sure the mrid used in JSON file (event) is in the query response. If this still doesn't run the simulation as desired, let's have a one-on-one meeting.

PREFIX r:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX c:  <http://iec.ch/TC57/CIM100#>
SELECT ?cimtype ?name ?bus1 ?bus2 ?id (group_concat(distinct ?phs;separator="") as ?phases) WHERE {
  SELECT ?cimtype ?name ?bus1 ?bus2 ?phs ?id WHERE {
 VALUES ?fdrid {"_AAE94E4A-2465-6F5E-37B1-3E72183A4E44"}  # 9500 node
 VALUES ?cimraw {c:LoadBreakSwitch c:Recloser c:Breaker}
 ?fdr c:IdentifiedObject.mRID ?fdrid.
 ?s r:type ?cimraw.
  bind(strafter(str(?cimraw),"#") as ?cimtype)
 ?s c:Equipment.EquipmentContainer ?fdr.
 ?s c:IdentifiedObject.name ?name.
 ?s c:IdentifiedObject.mRID ?id.
 ?t1 c:Terminal.ConductingEquipment ?s.
 ?t1 c:ACDCTerminal.sequenceNumber "1".
 ?t1 c:Terminal.ConnectivityNode ?cn1. 
 ?cn1 c:IdentifiedObject.name ?bus1.
 ?t2 c:Terminal.ConductingEquipment ?s.
 ?t2 c:ACDCTerminal.sequenceNumber "2".
 ?t2 c:Terminal.ConnectivityNode ?cn2. 
 ?cn2 c:IdentifiedObject.name ?bus2
	OPTIONAL {?swp c:SwitchPhase.Switch ?s.
 	?swp c:SwitchPhase.phaseSide1 ?phsraw.
   	bind(strafter(str(?phsraw),"SinglePhaseKind.") as ?phs) }
 } ORDER BY ?name ?phs
}
GROUP BY ?cimtype ?name ?bus1 ?bus2 ?id
ORDER BY ?cimtype ?name

**Reply by shpoudel** [reply]
Still, the query is not copied correctly. Please replace the first two lines of the query with this:

![image](https://user-images.githubusercontent.com/35777492/110817821-88878500-8241-11eb-80d6-2f5d65b7921f.png)

**Reply by Lakes-lead** [reply]
Hello @shpoudel the query is working now, but i am still getting a response attached. 
![Samerror](https://user-images.githubusercontent.com/26089614/110827716-589da900-8264-11eb-951c-bee6c4d14de4.PNG)

Whenever you are free, here is a link for the one on one (https://meet.google.com/dao-fthv-wvt)  Thank you.

**Reply by Lakes-lead** [reply]
@shpoudel You can use the calendly link to select a schedule convenient with you 
https://calendly.com/d/gq2p-v3fc/gridappsd-one-on-one

Thank you

**Reply by shpoudel** [reply]
@Lakes-lead Please send me your university email and I will try to find some time for us to meet.

**Reply by Lakes-lead** [reply]
oogundai@uncc.edu
Thanks
